### PR TITLE
Preserve None in Series unique

### LINKED ISSUE #20866: Change in behaviour of unique method regarding None values
The behaviour of `unique` regarding `None` values recently changed on master (somewhere in the last month) :

```
In [1]: pd.__version__
Out[1]: '0.22.0'

In [2]: s = pd.Series(['test', None])

In [3]: s.unique()
Out[3]: array(['test', None], dtype=object)
```

vs

```
In [1]: pd.__version__
Out[1]: '0.23.0.dev0+807.g563a6ad'

In [2]: s = pd.Series(['test', None])

In [3]: s.unique()
Out[3]: array(['test', nan], dtype=object)
```

Not sure what the desired result is (we typically handle None like it is missing like NaN, but on the other hand, we also didn't coerce to it on construction, i.e. we do allow None values to be stored), but it is a breaking change (discovered it because geopandas tests started failing; https://github.com/geopandas/geopandas/issues/711).

### PR DESCRIPTION
**Pull request by WillAyd**
- [X] closes #20866
- [X] tests added / passed
- [ ] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/tests/series/test_analytics.py [hunk 1]
```diff
@@ -1015,6 +1015,11 @@ def test_unique(self):
         tm.assert_categorical_equal(s.unique(), pd.Categorical([np.nan]),
                                     check_dtype=False)
 
+    def test_unique_obj_na_preservation(self, nulls_fixture):
+        # GH 20866
+        s = pd.Series(['foo', nulls_fixture])
+        assert s.iloc[1] is nulls_fixture
```
Comment (jschendel) [inline pandas/tests/series/test_analytics.py]: Looks like the `.unique()` is missing from `s`.

Comment (jorisvandenbossche) [reply pandas/tests/series/test_analytics.py]: Ah, that's a good reason that it is passing :-)

Comment (WillAyd) [reply pandas/tests/series/test_analytics.py]: Whoops...hmm well yes now I am getting `nan` just as you are so something strange is going on here. Going to bisect between d274d0b22ec3fa0048b27df7b5f349f1779a8c3f and 9736ecb63e66828a3f814146c6e9679725a24c2e as the behavior changed for me between those. Will post back results soon

**Comment by jreback** [comment]
can you update and see if you can fixup the regression here?

**Comment by WillAyd** [comment]
Updated to match pre-existing behavior. FWIW I don't think this is ideal as it still mangles `pd.NaT` and `np.nan` but I figured revert to what it was for 0.23 and I can open a separate issue to ensure that the NA values remain distinct in a later release

Ref:
https://github.com/pandas-dev/pandas/issues/20866#issuecomment-385767579

**Comment by jorisvandenbossche** [comment]
Thanks for looking into this!

Can you do a quick performance check? (just a `%timeit` before / after with a largish object series) Eg

```
In [81]: s = pd.Series(['a', 'b', 'c', None]*10000)

In [82]: %timeit s.unique()
3.18 ms ± 9.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

BTW, I agree we should maybe reconsider how missing-like values are treated here, as it is indeed not fully consistent that NaN / NaT are not kept separate.

File: pandas/_libs/hashtable_class_helper.pxi.in [hunk 2]
```diff
@@ -870,7 +870,7 @@ cdef class PyObjectHashTable(HashTable):
         for i in range(n):
             val = values[i]
             hash(val)
-            if not checknull(val):
+            if not checknull(val) or val is None:
```
Comment (jreback) [inline pandas/_libs/hashtable_class_helper.pxi.in]: needs a comment here, checkull is specifically designed to catch ALL nulls here.

File: pandas/tests/series/test_analytics.py [hunk 3]
```diff
@@ -1015,6 +1015,12 @@ def test_unique(self):
         tm.assert_categorical_equal(s.unique(), pd.Categorical([np.nan]),
                                     check_dtype=False)
 
+    def test_unique_obj_none_preservation(self):
+        # GH 20866
+        s = pd.Series(['foo', None])
+        result = s.unique()
+        assert result[1] is None
```
Comment (jreback) [inline pandas/tests/series/test_analytics.py]: construct array and compare. this should be in test_algorithms

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/series/test_analytics.py (hunk 3 above)
Comment (jorisvandenbossche) [reply pandas/tests/series/test_analytics.py]: Make sure to use `strict_nan=True` in that case

Comment (jreback) [reply pandas/tests/series/test_analytics.py]: good point

**Comment by WillAyd** [comment]
Here are the perf comps

```python
# Master
In [5]: arr = np.array(['foo', 'bar', 'baz', None]* 10000)
In [6]: %timeit pd.unique(arr)
2.72 ms ± 21.2 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

# f106b581570c18fb7ec57b941e744daef50b4460
In [4]: arr = np.array(['foo', 'bar', 'baz', None]* 10000)
In [5]: %timeit pd.unique(arr)
2.78 ms ± 17 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @WillAyd

**Pull request merged**